### PR TITLE
Replace setup.py with pyproject.toml and implement optional dependencies

### LINKED ISSUE #130: Install dependencies based on core, dev, doc or test use case
Currently everything in [requirements.txt](https://github.com/quaquel/EMAworkbench/blob/master/requirements.txt) is always installed no matter the use case. For essential usage (core) a lot less dependencies are needed than for development or docs generation. _sphinx_ and _nbsphinx_ (and _pyscaffold_?) are for example only needed when building docs. The _jupyter_ and _ipython_ stuff is also only needed in some cases.

I was thinking of using [Extras](https://peps.python.org/pep-0508/#extras) as defined in PEP 508.

Suggested options for PyPI:

- `pip install ema-workbench` installs the ema-workbench and core dependencies (numpy etc.)
- `pip install ema-workbench[jupyter]` installs all the Jupyter stuff to run in notebooks including parallel
- `pip install ema-workbench[dev]` installs everything needed for development and testing
- `pip install ema-workbench[docs]` installs everything for doc generation
- `pip install ema-workbench[all]` installs a superset of all of the above.

Combinations can be used, like `pip install ema-workbench[jupyter,dev]`. 

I need a bit of time to sort out how to structure all this, but if agreed this is useful I can create a draft PR.

### PR DESCRIPTION
**Pull request by EwoutH**
This PR does two things:
 - It migrates the project install commands and metadata from `setup.py` to `pyproject.toml`. It largely follows the [convention from setuptools](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html).
 - It moves the dependencies from `requirements.txt` to the `pyproject.toml` file, and thereby also defines a few sets of optional depenencies. Note that many dependencies that were defined in the `requirements.txt` are now optional.

This PR succeeds #151 and closes #130.

**To-do:**

 - Migrate to `pyproject.toml`:
   - [x] Project and Metadata 
   - [x] Dynamic version
   - [x] Dependencies and optional dependencies
   - [x] Version 
   - [x] Packaging
 - [x] Update CI to use appropriate install commands for each job
 - [x] Update installation docs
 - [x] Implement `"all"` extras which installs everything
 - [x] Test installing `"all"` extra on Linux and macOS (does Pythonnet error?)

The installation.rst docs are also updated to reflect the new options for installing extras.

### CODE REVIEW

**Comment by EwoutH** [comment]
@quaquel One thing still needs to be implemented, and I need a bit of help from you on that. That is the packaing, or what files/directories to include in the package. Currently there is [quite a large script](https://github.com/quaquel/EMAworkbench/blob/f11baaa47af4d041ac080d1ab96fd738a2b46190/setup.py#L27-L54) for, but I'm sure we could simplify that.

Could you tell me which folders and files need to be exactly included in the package? Then I will take an initial jab at implementing that in the pyproject.toml.

**Comment by EwoutH** [comment]
With a bit of reverse engineering I figured it out. :)

The reverse-wildcard trick is brand new and not even documented in Setuptools yet (I found it in a PR, and requested it to be documented in the [Data Files Support](https://setuptools.pypa.io/en/latest/userguide/datafiles.html) docs).

**Comment by EwoutH** [comment]
@quaquel Ready for review! Please also check the [updated installation instructions](https://github.com/quaquel/EMAworkbench/blob/pyproject/docs/source/installation.rst), and if the right dependencies are assigned to each of the extras.

**Comment by quaquel** [comment]
You might have to walk me through all these changes in person. A couple of things
1. In the past, I had bad experiences with the automated installation of dependencies. I have seen in brake quite a few anaconda installations. This might be from before `pyproject.toml` so it might no longer be a problem
2.  I am not sure whether we want to break down the dependencies to the degree you have done here. Yes, keeping a light installation can be useful in some cases, but it also creates a possible problem for users that might not fully understand all the configuration options that they have. Again, we can discuss the exact options in person.
3. I would not call `["pysd"]` `vensim` as is now done in the [updated installation instructions](https://github.com/quaquel/EMAworkbench/blob/a3747a235da7eaad78e54c44997c5e1dc3b27450/docs/source/installation.rst). Vensim itself only requires the DSS version and is really a different connector from pysd. I suggest just calling `["pysd"]` `pysd`.

**Comment by EwoutH** [comment]
Thanks for reviewing! I will look into your points, let's walk through them Monday.

Can we do 16:00 by the way? I won't be able to make 15:30.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+[tool.setuptools.dynamic]
+version = {attr = "ema_workbench.__version__"}
+
+[tool.setuptools.package-data]
+"ema_workbench.examples.data" = ["**"]
+"ema_workbench.examples.models" = ["**"]
+
+[project.optional-dependencies]
+recommended = ["ema_workbench[jupyter,dev]"]
```
Comment (EwoutH) [inline pyproject.toml:40]: Add graph

**Comment by EwoutH** [comment]
@quaquel Turns out both JPype and pythonnet crash on other OSes than Windows.

In the latest commit:
- Added `graph` to recommended extra
- Removed the connector dependencies from `all` extra
- Updated installation.rst docs

On my side it's ready, if you agree, I will clean it up and merge!

**Comment by EwoutH** [comment]
@quaquel If you approve I can cleanup and merge, and modify #170 to work with the new setup.

**Comment by quaquel** [comment]
looks good so go ahead

**Pull request merged**